로드맵 페이지 달력 뷰 구현

rudy3091 opened the pull request:
## 😀 제목

로드맵 페이지 달력 뷰 구현

## ⛅️ 내용

> 이 PR의 작업 요약

- 로드맵 페이지에서 버튼 인터랙션이 가능하도록 했습니다
- 화면이 좁아서 한달 단위로 이동 가능하도록 했던 것을 14일 단위로 이동하도록 했습니다
- 버튼에 아이콘을 안쓰고 그냥 `<`, `>`로 처리했는데, 이부분은 추가할 예정입니다
- 달력 관련 유틸함수를 몇개 구현했습니다

## 🎸특이사항

> 리뷰시 참고할만한 내용, 주의깊게 봐줬으면 하는 내용

오늘 코드에는 특이사항이 없었습니다
현재 에픽 삭제/수정에 관한 사항을 정해놓지 않은것 같은데 제가 임의로 구현해도 될까요?
또 컨트롤러 로직에 관해 테스트코드를 쓰려했는데 생각처럼 되지 않아서 방법을 알고계신 분은 공유해주시면 감사하겠습니다..

rudy3091 committed:
chore: yarn.lock 파일 최신화

모듈 설치에 따른 yarn.lock 파일 최신화

rudy3091 committed:
feat: RoadmapCalendar 컴포넌트 수정 [#18]

현재 일자를 기준으로 2주간의 날짜를 표시하도록 구현함

rudy3091 committed:
feat: calendar view 구조 설정 [#18]

디버그를 위해 border 색상으로 영역을 구분함

rudy3091 committed:
layout: style 속성 추가 [#18]

일자를 나타내는 label 중앙정렬 추가

rudy3091 committed:
feat: 유틸함수 getYMD 추가 [#18]

date 객체를 받아 그 date 의 연 월 일 정보를 반환하는 함수

rudy3091 committed:
feat: 일자를 바꿀수 있도록 구현 [#21]

좌우 버튼을 클릭하면 2주일 단위로 일자가 변경됨

rudy3091 committed:
refactor: 매직넘버 분리 [#21]

2주에 해당하는 14라는 수를 변수로 분리

rudy3091 committed:
feat: 날짜관련 유틸함수 2개 추가

윤년인지의 여부를 계산하는 isLeapYear
그 달의 마지막 일자를 계산하는 getLastDate

rudy3091 committed:
fix: getLastDate 유틸함수 문제 해결

fallthrough 문법 오류 수정

rudy3091 committed:
refactor: 매직넘버 분리

8일과 7일에 해당하는 숫자를 상수화

rudy3091 committed:
feat: 마지막 날짜를 넘어가는 일자 처리 [#21]

11월의 30일을 넘어가는 31, 32, 33 을 12월의 1, 2, 3일로 보여주도록 수정

rudy3091 committed:
layout: 디버그용으로 남겨둔 스타일 제거 [#21]

디버그용으로 남겨둔 DayColumn의 border 속성 제거

rudy3091 committed:
layout: 로드맵 페이지 간격 조정 [#18]

로드맵 페이지에서 달력에 해당하는 오른쪽 영역과 에픽 목록이 표시되는
왼쪽 영역의 위쪽 간격을 맞춤

rudy3091 committed:
feat: 에픽 삭제로직 추가

에픽을 삭제할 수 있도록 핸들러 함수 작성

rudy3091 committed:
refactor: 에픽 생성 함수 수정 [#15]

변수 projectName 을 projectId 로 수정
Date를 클라이언트에서 생성하고 이를 요청에 포함시키도록 수정

rudy3091 committed:
refactor: 필요없는 파일 제거

쓸일없는 sql 코드 제거

rudy3091 committed:
layout: 스타일 수정 [#12]

가로로 긴 에픽이름이 줄바꿈되지 않도록 수정

rudy3091 committed:
feat: RoadmapItem 컴포넌트 추가 [#22]

로드맵 항목의 막대 역할을 하는 컴포넌트

rudy3091 committed:
layout: 드래그 인터랙션을 위한 스타일 조정 [#21]

position absolute 속성을 줘서 막대가 쌓이는것을 방해하지 않도록 수정

ChanHoLee275 commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
 const RoadmapCalendar = () => {
   const [date, setDate] = useState(new Date());
+
+  const makeDayRow = (date: Date) => {
+    const { year, month } = getYMD(date);
+    const lastDate = getLastDate({ year, month });
+    const formerLastDate = getLastDate({ year, month: month - 1 });
+    const daysArray = new Array<number>();
+    [...Array(FRONT_HALF)]
+      .map((_, i) => date.getDate() - i)
+      .map((day) => (day < 0 ? formerLastDate + day : day))
+      .reverse()
+      .forEach((day: number) => daysArray.push(day));
+    [...Array(REAR_HALF)]
+      .map((_, i) => date.getDate() + i + 1)
+      .map((day) => (day >= lastDate ? day - lastDate : day))
+      .forEach((day: number) => daysArray.push(day));
+    return daysArray;
+  };
```
컴포넌트의 작동과는 크게 상관이 없는 함수여서 함수 밖에 선언해서 사용해도 괜찮을 것 같아요!

ChanHoLee275 commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
-        {Array.from(new Array(14), (_, i) => (
-          // TODO: key props를 index 값이 아닌 date 값과 연관한 값으로 변경
-          <div key={i}>{i + 1}</div>
+        {makeDayRow(date).map((day: number) => (
```
makeDayRow가 2번 사용되는 것 같은데, 이러면 변수로 만드는 것은 어떨까요?

haesoo-y commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
+
+const WEEK_OFFSET = 14;
+const FRONT_HALF = 8;
+const REAR_HALF = 7;
```
👍

haesoo-y commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
+    const newDate = new Date(year, month, day + WEEK_OFFSET);
+    setDate(newDate);
+  };
+
```
위 아래 함수가 거의 같네요. changeYMD 라는 함수로 만들어서 인자에 따라 WEEK_OFFSET에 +1 또는 -1 을 곱하는 방식을 써도 좋을 것 같습니다.

haesoo-y commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
         ))}
       </S.DaysWrapper>
+      <S.DayColumnWrapper>
+        {makeDayRow(date).map((day: number) => (
```
찬호님 의견과 비슷한데  `makeDayRow`를 외부에 선언하고 컴포넌트에서는 `useMemo`로 불러와서 재활용하면 좋겠네요.

haesoo-y commented on client/src/lib/utils/date.ts:
```diff
@@ -8,3 +8,28 @@ export const makeEpicWithDate = (epicWithString: EpicWithString): EpicType => {
+    default:
+      return 31;
+  }
+};
```
거의 알고리즘 문제를 푸신 수준인데요??? 놀랐습니다 😱 
매직넘버 같은게 조금 아쉽네요. 해결하지 않아도 될 것 같지만 만일 해결한다면 어떻게 해결할 수 있을까요... 
```js
const monthThirtyOne = [3, 5, 8, 10]
const monthThirty = [0, 2, 4, 6, 9, 11]
if (month in monthThirtyOne) return 31;
if (month in monthThirty) return 30;
return isLeapYear(year) ? 29 : 28;
```
스위치문 대신 조건문으로 이렇게 되려나요 ㅋㅋ 참고만 해주세요

ingong replied:
ㅋㅋㅋㅋㅋㅋ윤년을 계산하시다니 놀랍습니다....!

ingong commented on server/src/Epics/Epics.controller.ts:
```diff
@@ -74,3 +74,24 @@ export const createEpic = async (req: Request, res: Response) => {
+ */
+export const deleteEpicById = async (req: Request, res: Response) => {
+  try {
+    await getRepository(Epics)
```
저는 queryBuilder 좋아요... 👍

ingong commented on yarn.lock:
```diff
@@ -4406,6 +4418,14 @@ builtin-status-codes@^3.0.0:
   resolved "https://registry.yarnpkg.com/builtin-status-codes/-/builtin-status-codes-3.0.0.tgz#85982878e21b98e1c66425e03d0174788f569ee8"
   integrity sha1-hZgoeOIbmOHGZCXgPQF0eI9Wnug=
 
+busboy@^0.2.11:
```
생각보다 수정이 많네요. 감안하고 가야겠죠...?

ingong commented on client/src/lib/utils/date.ts:
```diff
@@ -8,3 +8,28 @@ export const makeEpicWithDate = (epicWithString: EpicWithString): EpicType => {
   };
 };
+
+export const getYMD = (date: Date) => {
```
함수명 좋아요....!

ingong commented on client/src/components/RoadmapItem/index.tsx:
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import S from './style';
+
+const RoadmapItem = () => {
```
맞아요. 이렇게 Item 을 결국 분리해서 책임을 나눠줘야할 것 같더라구요!

ingong reviewed (commented):
졸려요...학교 과제하고 이제 리뷰 답니다...흑

ingong commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -1,23 +1,64 @@
+    const lastDate = getLastDate({ year, month });
+    const formerLastDate = getLastDate({ year, month: month - 1 });
+    const daysArray = new Array<number>();
+    [...Array(FRONT_HALF)]
```
혹시 REAR_HALF 와 FRONT_HALF 의 의미가 뒤, 앞의 의미인가요? 내일 설명듣고싶어요ㅎㅎ

rudy3091 committed:
refactor: 유틸파일 관심사 분리

makeEpicWithDate 라는 함수는 date와 관심사가 다른 함수이므로 epic.ts 에 분리함

rudy3091 committed:
refactor: 달력 뷰에 대한 리팩토링 수행

- 컴포넌트 내부에 선언되어 있던 makeDayRow 함수를 utils로 이동
- makeDayRow 함수는 상수 FRONT_HALF, REAR_HALF 에 대한 참조를 가져 해당
  상수 역시 같이 이동시켜줌
- 오늘 날짜가 row의 중간에 위치하도록 수치 조정

rudy3091 committed:
Merge branch 'develop' into feature/roadmap-calendar
